File: Y2018/Day09.hs
import Control.Monad
import Data.Function
import Data.List
import Data.Array hiding ((!))
import qualified Data.Array as Array ((!))
import qualified Data.Foldable as Foldable


addN :: (Num a) => Int -> a -> Array Int a -> Array Int a
addN i n l = l // [(i, (l Array.! i) + n)]

data Cycle a = Cycle ([a], Int)

instance Foldable Cycle where
    foldr f z x = foldr f z (header' x)

header (l, h) = let (l2, l1) = splitAt h l in l1 ++ l2
header' (Cycle x) = header x

nIndex (Cycle (l, h)) i = i `mod` (length l)

(!) :: Cycle a -> Int -> a
x ! i = (header' x) !! (nIndex x i)

move :: Int -> Cycle a -> Cycle a
move n (Cycle (l, h)) = Cycle (l, (h + n) `mod` (length l))

-- Insert between head and tail, and make it head
insertH :: a -> Cycle a -> Cycle a
insertH v x = Cycle (v : (header' x), 0)

deleteH :: Cycle a -> Cycle a
deleteH = snd . deleteFindH

deleteFindH :: Cycle a -> (a, Cycle a)
deleteFindH x = let (hd:tl) = header' x in (hd, Cycle (tl, 0))


initCircle :: Cycle Int
initCircle = Cycle ([0], 0)

initScore :: Int -> Array Int Int
initScore n = listArray (0, n) (repeat 0)

putMarble t c = ((insertH t) . (move 2)) c

putMarble23 nPlayer t c s = (c', s') where
    player = t `mod` nPlayer
    p1 = t
    (p2, c') = (deleteFindH . (move (-7))) c
    s' = addN player (p1 + p2) s

game nPlayer = (initScore nPlayer) : turn nPlayer 1 initCircle (initScore nPlayer) where
    turn nP t c s
        | (t `mod` 23) /= 0 = let c' = putMarble t c in s : turn nP (t+1) c' s
        | otherwise         = let (c', s') = putMarble23 nP t c s in s' : turn nP (t+1) c' s'

getInput :: String -> (Int, Int)
getInput line = (read $ w !! 0, read $ w !! 6) where w = words line

solve1 :: Int -> Int -> Int
solve1 nPlayer turn = maximum scores where
    scores = map snd (assocs $ game nPlayer !! turn)

-- This can't be computed, since it is not implemented as doubly linked lists
solve2 :: Int -> Int -> Int
solve2 nPlayer turn = solve1 nPlayer (turn * 100)

main' = do
    (nPlayer, turn) <- getInput <$> getLine
    putStrLn $ show $ solve1 nPlayer turn
    putStrLn "Solution 2 can't be computed, since it is not implemented as doubly linked lists."
    putStrLn $ show $ solve2 nPlayer turn

File: Y2018/Day09.py
import re


class Doubly:
    def __init__(self, value, prev=None, next=None):
        self.value = value
        self.prev = prev or self
        self.next = next or self

    def move(self, n):
        curr = self
        for i in range(abs(n)):
            if n < 0:
                curr = curr.prev
            else:
                curr = curr.next
        return curr

    def insert(self, v):
        prev = self.prev
        new_node = Doubly(v, prev, self)
        prev.next = new_node
        self.prev = new_node
        return new_node

    # Make sure 'del' this too
    def delete(self):
        self.prev.next = self.next
        self.next.prev = self.prev
        return self.value, self.next

def put_marble(t, c):
    return c.move(2).insert(t)

def put_marble_23(n_player, t, c, s):
    player = t % n_player
    p1 = t
    (p2, nc) = c.move(-7).delete()
    del c
    s[player] += p1 + p2
    return nc, s

def game(n_player, max_turn):
    c = Doubly(0)
    s = [0 for i in range(n_player + 1)]
    for t in range(1, max_turn + 1):
        if t % 23 != 0:
            c = put_marble(t, c)
        else:
            (c, s) = put_marble_23(n_player, t, c, s)
    return s

def solve1(n_player, turn):
    return max(game(n_player, turn))

def solve2(n_player, turn):
    return max(game(n_player, turn * 100))

with open('09.txt') as f:
    line = f.read()
    [n_player, turn] = [int(x) for x in re.search(r'(\d+)[^\d]*(\d+).*$', line).groups()]
    print(solve1(n_player, turn))
    print(solve2(n_player, turn))
